src/System.Windows.Forms.Primitives/src/Interop/WinFormsComWrappers.cs

```
@@ -50,10 +50,27 @@ protected override object CreateObject(IntPtr externalComObject, CreateObjectFla
             Debug.Assert(flags == CreateObjectFlags.UniqueInstance);
 
             Guid pictureIID = IID.IPicture;
-            int hr = Marshal.QueryInterface(externalComObject, ref pictureIID, out IntPtr comObject);
+            int hr = Marshal.QueryInterface(externalComObject, ref pictureIID, out IntPtr pictureComObject);
             if (hr == S_OK)
             {
-                return new PictureWrapper(comObject);
+                Marshal.Release(externalComObject);
+                return new PictureWrapper(pictureComObject);
+            }
+
+            Guid fileOpenDialogIID = IID.IFileOpenDialog;
+            hr = Marshal.QueryInterface(externalComObject, ref fileOpenDialogIID, out IntPtr fileOpenDialogComObject);
+            if (hr == S_OK)
+            {
+                Marshal.Release(externalComObject);
+                return new FileOpenDialogWrapper(fileOpenDialogComObject);
+            }
+
+            Guid shellItemIID = IID.IShellItem;
+            hr = Marshal.QueryInterface(externalComObject, ref shellItemIID, out IntPtr shellItemComObject);
+            if (hr == S_OK)
+            {
+                Marshal.Release(externalComObject);
+                return new ShellItemWrapper(shellItemComObject);
             }
 
             throw new NotImplementedException();
```

Would we not want to call `Marshal.Release(externalComObject);` before throwing the exception?
Should we do this:
```cs
try
{
    // if (..) return ...
}
finally 
{
    Marshal.Release(externalComObject);
}

throw new NotImplementedException();
```

---

I think we should not do that. If we reach this path, that's mean something within WinForms pass to internal ComWrappers borked value. That should not happens. That means serious bug for me which should be fixed.

---

Indeed we have a bug, we thrown and the app crashed... But would we not leak the COM object in this situation? I'm genuinely curious.

---

We can leak COM object if it is created out-of-process. That may happens only if we work with WebBrowser or DataObject probably. All of them are problematic even for this approach and better discuss that problem when we implement this parts.

What if COM object would be passed by the user(this is how I think we end-up in this hypotetical scenario) if we release his object during exception(our bug) he maybe unhappy since that's leave him without workaround. I'm more in the camp let it fail since that means major screwup on our side.